src/DotNet.Status.Web/Controllers/AzurePipelinesController.cs

```diff
@@ -27,44 +26,33 @@ namespace DotNet.Status.Web.Controllers
     public class AzurePipelinesController : ControllerBase
     {
         private readonly IGitHubApplicationClientFactory _gitHubApplicationClientFactory;
-        private readonly IAzureDevOpsClientFactory _azureDevOpsClientFactory;
+        private readonly IClientFactory<IAzureDevOpsClient> _azureDevOpsClientFactory;
         private readonly IOptionsSnapshot<BuildMonitorOptions> _options;
         private readonly ILogger<AzurePipelinesController> _logger;
-        private readonly Lazy<IAzureDevOpsClient> _clientLazy;
-        private readonly Lazy<Task<Dictionary<string, string>>> _projectMapping;
 
         public AzurePipelinesController(
             IGitHubApplicationClientFactory gitHubApplicationClientFactory,
-            IAzureDevOpsClientFactory azureDevOpsClientFactory,
+            IClientFactory<IAzureDevOpsClient> azureDevOpsClientFactory,
             IOptionsSnapshot<BuildMonitorOptions> options,
             ILogger<AzurePipelinesController> logger)
         {
             _gitHubApplicationClientFactory = gitHubApplicationClientFactory;
             _azureDevOpsClientFactory = azureDevOpsClientFactory;
             _options = options;
             _logger = logger;
-            _clientLazy = new Lazy<IAzureDevOpsClient>(BuildAzureDevOpsClient);
-            _projectMapping = new Lazy<Task<Dictionary<string,string>>>(GetProjectMappingInternal);
         }
 
-        private IAzureDevOpsClient Client => _clientLazy.Value;
-
-        private IAzureDevOpsClient BuildAzureDevOpsClient()
+        private Reference<IAzureDevOpsClient> GetAzureDevOpsClient()
         {
             BuildMonitorOptions.AzurePipelinesOptions o = _options.Value.Monitor;
-            return _azureDevOpsClientFactory.CreateAzureDevOpsClient(o.BaseUrl, o.Organization, o.MaxParallelRequests, o.AccessToken);
-        }
-
-        private async Task<Dictionary<string, string>> GetProjectMappingInternal()
-        {
-            var projects = await Client.ListProjectsAsync();
-            return projects.ToDictionary(p => p.Id, p => p.Name);
+            return _azureDevOpsClientFactory.GetClient($"build-monitor/{o.Organization}");
         }
 
-        private async Task<string> GetProjectNameAsync(string id)
+        private async Task<string> GetProjectNameAsync(IAzureDevOpsClient client, string id)
```

Comment:
Should this function just be moved the the client so we don't have to pass the client as a parameter?

Reply:
sure